### DISCUSSION TITLE
No DB shards could be opened - error after migrating to new server

### DISCUSSION THREAD

**Discussion by PaulMcF1987** [Budibase/budibase#6472] [Help]
I have moved to a new server and tried to take my Budibase instance with me.

I followed the steps [here](https://docs.docker.com/desktop/backup-and-restore/)

It appears that the bbapps image wont start

The error messages I am getting:

`[Tenant: default] [Migration: user_email_view_casing] [DB: global-db] Error:  {`
`  error: 'internal_server_error',`
 ` reason: 'No DB shards could be opened.',`
 ` ref: 179462285,`
 ` status: 500,`
 ` name: 'internal_server_error',`
 ` message: 'No DB shards could be opened.',`
 ` stack: 'Error\n' +`
 `   '    at generateErrorFromResponse (/app/node_modules/pouchdb/lib/index.js:694:18)\n' +`
 `   '    at /app/node_modules/pouchdb/lib/index.js:6816:19\n' +`
 `   '    at processTicksAndRejections (internal/process/task_queues.js:95:5)'`
`}`
`bb-alert: Error performing migrations. Exiting. : {"error":"internal_server_error","reason":"No DB shards could be opened.","ref":179462285,"status":500,"name":"internal_server_error","message":"No DB shards could be opened.","stack":"Error\n    at generateErrorFromResponse (/app/node_modules/pouchdb/lib/index.js:694:18)\n    at /app/node_modules/pouchdb/lib/index.js:6816:19\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)"}`

**Comment by mike12345567** [comment]
Hi @PaulMcF1987 - there's a few things to check here, from the errors it looks like CouchDB hasn't correctly loaded, it is unable to access to shards within the volume.

Things to look into:
1. How you exported your image/volumes, if you pulled a new image for Couch this could cause problems due to a version mismatch.
2. Check your environment variables and configuration, if your CouchDB password has changed or anything along those lines, you'll run into issues with the instance being unable to re-open the files on the volume.
3. You can try connecting to CouchDB directly and checking whether any of your databases have been correctly imported, you can do this by going to `http://<your host>:10000/db/_utils`, signing in with your CouchDB password and looking through to see if any of your app databases can be accessed.

**Reply by PaulMcF1987** [reply]
@mike12345567 
I used the load feature to load the .tar file that I had exported, however, when I started up budibase, It seemed to pull a new image from the docker repositary.
How Do I run budibase using only the files that I have imported via the .tar?
Also... goingforward, if budibase updates, will the problems persist?

**Reply by PaulMcF1987** [reply]
@mike12345567 
Tried step 3 - `http://<your host>:10000/db/_utils`

Takes me to the `Houston we have a problem` page

**Reply by PaulMcF1987** [reply]
I have tried to change the names of the Image files in the YAML file to the names of the images that I have exported but it still pulls the images from the docker repository and ignores what is in the YAML file

**Reply by mike12345567** [reply]
Hi @PaulMcF1987 Sorry, please try going to `http://<your host>:10000/db/_utils/#login` to get to the database.

In terms of loading the images, you'll need to load the images on the new machine, with something like `docker <image name> load -i image.tar`. The Budibase images in this case shouldn't matter, it is more than likely an environment variable difference than an image version difference - you shouldn't really need to load the images, you should be able to just pull them, but to load a CouchDB volume directly you'll need to make sure the versions are identical and loading the Couch image is an easy way to confirm this.

**Reply by PaulMcF1987** [reply]
@mike12345567 I have removed all instances of budibase and am trying to reinstall the images
`docker <image name> load -i image.tar` is what I used, however, this only adds the images to the machine, it does not launch the containers as you will see from the below pic, the images are there but highlighted as unused.

I have spent the last 2-3 hours looking for a way to use these images to launch the containers with no joy.

I tried changing the image name in the YAML file to match the local image but it didnt work, it pulled images from the docker repositary

![image](https://user-images.githubusercontent.com/75368133/175977422-0828bb48-4515-4641-8337-d17f48089de3.png)

The ENV file shouldnt have any differences as I copied the YAML and ENV straight over from the old server, I made no changes.

**Reply by PaulMcF1987** [reply]
@mike12345567 I have managed to get it up and running  using the exported volumes and containers. however, it looks like a new set up

![image](https://user-images.githubusercontent.com/75368133/175999471-4a797fb1-01f7-489d-9f89-bb4562699f08.png)

I tried visiting `http://<your host>:10000/db/_utils/#login` but my user details were not recognised

**Reply by mike12345567** [reply]
Hi @PaulMcF1987 - if your CouchDB credentials are not recognised by the DB login then it sounds like a new system has been generated, either the volumes haven't exported correctly, or your environment variables have changed in some manner. If you are unable to login to the database using the credentials set in your `.env` file then it hasn't picked up your environment variables file correctly - this needs to be corrected before the volumes will be able to import correctly.

**Reply by PaulMcF1987** [reply]
@mike12345567 
I dont know what Im looking for

My env file:
`# Use the main port in the builder for your self hosting URL, e.g. localhost:10000`
`MAIN_PORT=10000`

`# This section contains all secrets pertaining to the system`
`# These should be updated`
`JWT_SECRET=Same_Secret_Key_as _used_on_other_server`
`MINIO_ACCESS_KEY=Same_Access_Key_as _used_on_other_server`
`MINIO_SECRET_KEY=Same_Secret_Key_as _used_on_other_server`
`COUCH_DB_PASSWORD=Same_Password_as _used_on_other_server`
`COUCH_DB_USER=Same_user_as _used_on_other_server`
`REDIS_PASSWORD=Same_password_as _used_on_other_server`
`INTERNAL_API_KEY=Same_api_Key_as _used_on_other_server`

`# This section contains variables that do not need to be altered under normal circumstances`
`APP_PORT=9900`
`WORKER_PORT=9901`
`MINIO_PORT=9902`
`COUCH_DB_PORT=9903`
`REDIS_PORT=9904`
`WATCHTOWER_PORT=9905`
`BUDIBASE_ENVIRONMENT=PRODUCTION`

**Reply by PaulMcF1987** [reply]
I am back to square 1 with `houston we have a problem`
The error has changed slighty to

`error: 'internal_server_error',`
  `reason: 'No DB shards could be opened.',`
  `ref: 179462285,`
  `status: 500,`
  `name: 'internal_server_error',`
  `message: 'No DB shards could be opened.',`
  `stack: 'Error\n' +`
`    '    at generateErrorFromResponse (/app/node_modules/pouchdb/lib/index.js:694:18)\n' +`
`    '    at /app/node_modules/pouchdb/lib/index.js:6816:19\n' +`
`    '    at processTicksAndRejections (internal/process/task_queues.js:95:5)'`
`}`
`[Tenant: default] [Migration: user_email_view_casing] [DB: global-db] Error:  TypeError: Cannot read property 'user_email_view_casing' of undefined`
    `at /app/node_modules/@budibase/backend-core/src/migrations/index.js:55:16`
    `at processTicksAndRejections (internal/process/task_queues.js:95:5)`
    `at async exports.doWithDB (/app/node_modules/@budibase/backend-core/src/db/index.js:62:12)`
    `at async runMigration (/app/node_modules/@budibase/backend-core/src/migrations/index.js:50:5)`
    `at async internal (/app/node_modules/@budibase/backend-core/src/context/index.js:90:14)`
    `at async /app/node_modules/@budibase/backend-core/src/context/FunctionContext.js:23:22`
    `at async exports.runMigrations (/app/node_modules/@budibase/backend-core/src/migrations/index.js:113:7)`
`bb-alert: Error performing migrations. Exiting. : {"stack":"TypeError: Cannot read property 'user_email_view_casing' of undefined\n    at /app/node_modules/@budibase/backend-core/src/migrations/index.js:55:16\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async exports.doWithDB (/app/node_modules/@budibase/backend-core/src/db/index.js:62:12)\n    at async runMigration (/app/node_modules/@budibase/backend-core/src/migrations/index.js:50:5)\n    at async internal (/app/node_modules/@budibase/backend-core/src/context/index.js:90:14)\n    at async /app/node_modules/@budibase/backend-core/src/context/FunctionContext.js:23:22\n    at async exports.runMigrations (/app/node_modules/@budibase/backend-core/src/migrations/index.js:113:7)","message":"Cannot read property 'user_email_view_casing' of undefined"}`

**Reply by PaulMcF1987** [reply]
@mike12345567 the problem seems to be communicating with the couchDB server. 
It seems to be the couchDB volume pulled from the old server. I have added all the volumes indevidually and restarted the container and its only when I add CouchDB volume that the problems start

rather than copying all of the volume for CouchDB over, is there a certain part that I can copy over to get the user table and the apps?

**Reply by mike12345567** [reply]
Hi @PaulMcF1987 - right now there isn't an easy way to export all users from a Budibase system and move them to somewhere new at the application level, this is something we plan to work on. You can export all of the apps individually and import them, however this isn't great if you have a lot of apps since they all need to be re-published as well.

This is something we will be looking at more in the future, we've thought about adding a full export option to the Budibase CLI to help with this, make it easier to carry out jobs like this.

Importing the volume to a new CouchDB container should work correctly, the main reasons that it wouldn't could be that the image isn't the identical couch one that was used for the old install, the environment configuration differs slightly or the image was started without the volume mounted correctly initially.

**Reply by PaulMcF1987** [reply]
@mike12345567 I used budibase a few months ago and removed it in favour of a similar application. I then decided to come back to budibase and had a similar problem in that I could not get it to work. It turned out that the reason that it would not work is that the volume was still in place from the previous install. Once I deleted the volume it worked fine.

So based on what you have said above, is it only the couchDB image that I need to copy over? the rest can be done with fresh images from the repository?
What would that then mean for updates etc?

If I Export the apps and import them on the new server... This wont import the users? which means that I will need to set all the users up with new login details?

**Reply by mike12345567** [reply]
Yes it should only be the CouchDB image that has issues I believe, the Budibase images are fairly compatible with each other, and can't cause issues like you're experiencing, it will realise version differences and update your apps as required - however low level exports like volumes etc can have weird effects on CouchDB when there are differences. For updates you should be fine once you get it up and running, its just about getting the database up and running initially.

On the export apps, we've not implemented a method of exporting/importing users to a new system yet because of the complexity of it - this is something we do want to fix, in general we want to make it easier to backup Budibase, export/import to new systems etc etc. This is something we're aware of and is in the backlog but as of yet we've not done this.

If you get the volume export to work you will get everything, users, apps, everything, will all import across correctly, but getting the environments identical can be a little difficult - again this is something that we want to work on, make an easy volume export/import method using the Budibase CLI - but again this is something on the backlog to be looked at in the future.

**Reply by PaulMcF1987** [reply]
@mike12345567 I have tried every possible way I can think of, adding the database volumes first and then pulling the images, pulling the images first and then the volumes... I get the same outcome every time
There shouldnt be any issue with the ENV as it is the exact same ENV that was used on the previous server

The only reason I had to upgrade the server was because since adding Budibase, the Swap Memory has been filling up, causing the CPU to spike and crashing the server

And it appears that changing server is not helping either. I have managed to migrate all of my other containers with no issues, the volumes paired up perfectly and I have been able to pick up where I left off.

Is there a better way to transfer the volumes over that might work better? As when I use the volume from the old server I cannot access anything I get the `houston we have a problem` I get the same message on `ip:10000/db/_utils/#login`

I have used the images from my old server which I exported in a tar file.

I am struggling to understand why this process is so difficult. all other containers transferred over without any difficulties and there were some complex containers such as open project but it works fine

**Reply by mike12345567** [reply]
Hi @PaulMcF1987 - as a team we've discussed this, it is a little difficult right now, its hard for me to know whats going wrong right now as there are so many variables that could have changed which have caused the problem.

We plan to update the Budibase CLI in the near future to carry out the process using CouchDB replications as this should avoid the need to export/import volumes - I'm not sure exactly why you're having this trouble but hopefully a tool like that should simplify the process significantly and automate it to a greater extent. All being well we will carry out this work in the near future.

It is quite complicated, but a way that avoids the need to volume export/import is to use CouchDB replications directly, if your old system has access to the new one you can try using Fauxton (The DB UI available on `http://ip:10000/db/_utils/#login`) to configure a a replication of each of the databases from the old system to the new one, [documentation here](https://couchdb.apache.org/fauxton-visual-guide/index.html#replication).

As I say we will be simplifying this in the near future with an automated tool built into the CLI.

**Reply by PaulMcF1987** [reply]
@mike12345567 I wish I had time to wait until the near future lol. I really needed this active yesterday as I am currently paying for 2 servers until I get this sorted

I will have a look at the above. however, that would mean setting up a new DB and working from that which would likely still mean giving all users new log in credentials which is a massive amount of work.

Is there no way to do a DB Dump with CouchDB?

**Reply by shogunpurple** [reply]
@PaulMcF1987 if you are having trouble with migrating the docker volumes, you may be able to use this tool - it depends how many apps you have, but if you only have a few databases it should do the trick:
https://github.com/danielebailo/couchdb-dump

After doing that, you will need to migrate over your MinIO files. You should be able to do this using the minIO CLI:
https://docs.min.io/docs/minio-client-complete-guide.html

You can download the files from your existing minIO cluster, and upload them to your new one. After migrating both couch and MinIO, your apps and users should be preserved.

**Reply by PaulMcF1987** [reply]
@shogunpurple will this work if the container is already made? or will I need to remove the container, run the above and then recreate the container once I have transferred the data?

**Reply by PaulMcF1987** [reply]
Out of curiousity, could it be related to the working directory?
On the old server, the files were located in the directory /budibase
On the new server, due to the way the memory is allocated I have had to set up the directory in /var/lib/docker/budibase

**Reply by shogunpurple** [reply]
> @shogunpurple will this work if the container is already made? or will I need to remove the container, run the above and then recreate the container once I have transferred the data?

Yes - the replication happens over HTTP, so if you have an instance of budibase running on each 2 server, you can perform the replication between the old one and the new one. You won't need to stop/start/restart anything at all.

**Reply by PaulMcF1987** [reply]
@mike12345567  @shogunpurple  I have tried the replication method but this is not working either

![image](https://user-images.githubusercontent.com/75368133/176219959-828f9de0-a6e6-4d49-81a0-d2faf238e896.png)

Does this matter?

![image](https://user-images.githubusercontent.com/75368133/176220801-8375f09c-886f-464a-9eee-5f7790d96748.png)


---------------------------------------------------------------------------------------------------------------------------------

I copied the old DB files and moved them to the new server and logged in to the back end and the DB files are there but apparently not accessible

![image](https://user-images.githubusercontent.com/75368133/176225181-0d3ae84a-b23a-4ef7-a91f-1559860d9548.png)

**Comment by mike12345567** [comment]
Hi @PaulMcF1987 - just to let you know/anyone else looking at this issue that there is a WIP CLI update which will be released soon to provide a better experience for backing up a whole Budibase installation in self host, as well as importing this backup to the same, or a new environment.

https://github.com/Budibase/budibase/pull/6584
